src/server/master_server/chunk_server_manager.cc

```diff
@@ -1,69 +1,304 @@
   }
 
-  return server;
+  return {server, server_lock};
+}
+
+void ChunkServerManager::UpdateChunkServer(
+    const protos::ChunkServerLocation& server_location,
+    const uint32_t& available_disk_mb,
+    const absl::flat_hash_set<std::string>& chunks_to_add,
+    const absl::flat_hash_set<std::string>& chunks_to_remove) {
+  auto result = this->chunk_servers_map_.find(server_location);
+
+  // If not found, nothing to update
+  if (result == this->chunk_servers_map_.end()) {
+    return;
+  }
+
+  // We need to update this chunk server. First remove the chunks that needs to
+  // be removed and add the new ones.
+  auto chunk_server = result->second;
+
+  // Lock guard scope
+  {
+    // We could be adding to chunkserver during allocation. Acquire lock here
+    // before adding/removing chunks from the chunkserver.
+    absl::WriterMutexLock chunk_server_write_lock_guard(
+        chunk_servers_lock_map_[chunk_server->location()].get());
+
+    // Remove chunks
+    if (!chunks_to_remove.empty()) {
+      for (auto iterator = chunk_server->stored_chunk_handles().begin();
+           iterator != chunk_server->stored_chunk_handles().end();) {
+        auto current_chunk_handle = *iterator;
+
+        if (chunks_to_remove.contains(current_chunk_handle)) {
+          // Remove this chunk from chunkserver
+          // Returns iterator to next element
+          iterator =
+              chunk_server->mutable_stored_chunk_handles()->erase(iterator);
+          // Remove this chunk server from the chunk handle location map.
+          this->chunk_locations_map_[current_chunk_handle].erase(
+              chunk_server->location());
+        } else {
+          ++iterator;
+        }
+      }
+    }
+
+    // Add chunks
+    for (auto iterator = chunks_to_add.begin(); iterator != chunks_to_add.end();
+         ++iterator) {
+      auto current_chunk_handle = *iterator;
+      chunk_server->add_stored_chunk_handles(current_chunk_handle);
+      // Add to chunk location map
+      this->chunk_locations_map_[current_chunk_handle].insert(
+          chunk_server->location());
+    }
+  }
+
+  // We drop the initial write lock before getting the priority list lock,
+  // because the priority list lock is gotten in chunk allocation before
+  // requesting chunkserver write lock. This is to avoid deadlock when we are
+  // running update while allocation is also going on. And for better
+  // performance don't want to block all allocations before doing update,
+  // that why i'm getting the priority list lock here.
+
+  // Taking this lock here, since an allocation could
+  // be resorting the priority list as we update the chunkserver disk.
+  absl::MutexLock chunk_servers_priority_list_lock_guard(
+      &chunk_servers_priority_list_lock_);
+
+  absl::WriterMutexLock chunk_server_write_lock_guard(
+      chunk_servers_lock_map_[chunk_server->location()].get());
+
+  // Update available disk.
+  chunk_server->set_available_disk_mb(available_disk_mb);
```

Comment:
are we assuming caller will always make the right arguments?

It is possible that 
1. after remove chunks and add chunks operations, the chunk server now has more chunk handles trying to be stored than allowed in either "available_disk_mb" or configuration's minimum disk space requirement
2. we are trying to reduce the chunks stored here, thus available_disk_mb is smaller than it was before; however, because of race condition between the last lock scope and here, a race condition makes another thread allocate  more chunks to the chunk sever, violating what "available_disk_mb" is supposed to do or modify, leaving inconsistencies

Reply:
Yes the arguments will be correct since this is called by the chunkservermgr service.

About:
1. You are right. Thought about this but was like this will still happen anyway, but at a reduced rate. Because there is a diff between the time the disk actually changed on the chunkserver and when we get the update and apply. Was trying to prioritize chunkserver allocation here since it is done in foreground and this is done in the background.
2.  available disk is not always smaller in fact I will say most of the time it will either be the same or larger.

All your points are correct, was trying to optimize for chunk allocation here. But will fix by acquiring the priority list lock first, before doing any update.